Framework/PythonInterface/plugins/algorithms/fractional_indexing.py

```diff
@@ -0,0 +1,302 @@
+            int -- k is the number of clusters used
+    """
+    if k is not None:
+        centroids = kmeans_plus_plus(qs, k)
+        _, clusters = kmeans2(qs, centroids, minit='matrix')
+        if len(set(clusters)) != k:
+            raise ValueError("Could not group the satellite reflections "
+                             "into {} clusters. Please check that you have "
+                             "at least {} satellites.".format(k,k))
+    else:
+        clusters = hcluster.fclusterdata(qs, threshold, criterion="distance")
+    return clusters, len(set(clusters))
+
+
+def find_q_vectors(nuclear_hkls, sats_hkls):
+    """Find the q vector between the nuclear HKLs and the satellite peaks
+
+    Given a list of HKL positions and a list of fractional HKL positions of
+    satellite peaks, find the difference between each satellite and its nearest
+    integer HKL.
+
+    :param nuclear_hkls: the positions of integer HKL peaks.
+    :param sats_hkl: the positions of fractional "satellite" HKL peaks.
+    :returns: np.ndarray -- array of q vectors.
+    """
+    peak_map = KDTree(nuclear_hkls)
+    qs = []
+    for sat in sats_hkls:
+        distance, index = peak_map.query(sat, k=1)
+        if distance > 2:
+            # peak to far away from satellite ignore
+            continue
+        nearest_peak = nuclear_hkls[index]
+        qs.append(sat - nearest_peak)
+    return np.array(qs)
+
+
+def find_nearest_integer_peaks(nuclear_hkls, sat_hkls):
+    """Find the nearest integer peak for each fractional peak
+
+    This will perform a spatial search to find the intger peak which is nearest the fractional satellite peak.
+
+    :param nuclear_hkls: the hkl poistions of each of the nuclear peaks.
+    :param sat_hkls: the fractional hkl position of the satellite peaks.
+    :returns: np.ndarray -- 2D array of HKL integer values for each fractional peak
+    """
+    peak_map = KDTree(nuclear_hkls)
+    hkls = []
+    for sat in sat_hkls:
+        distance, index = peak_map.query(sat, k=1)
+        nearest_peak = nuclear_hkls[index]
+        hkls.append(nearest_peak)
+    return np.array(hkls)
+
+
+def average_clusters(qs, clusters):
+    """Compute the centroid of each cluster of vectors
+
+    :param qs: the q vectors to find the centroids of
+    :param clusters: the cluster index that each q belongs to
+    :return: an ndarray of cluster centroids
+    """
+    averaged_qs = []
+    for cluster_index in set(clusters):
+        averaged_qs.append(np.mean(qs[clusters == cluster_index], axis=0))
+    return np.array(averaged_qs)
+
+
+def get_hkls(peaks_workspace):
+    """Get the H, K, and L columns from a PeaksWorkspace.
+
+    :param peaks_workspace: the peaks workspace to extract HKL values from
+    :return: 2D numpy array of HKL values.
+    """
+    return np.array([np.array([peak['h'], peak['k'], peak['l']])
+                     for peak in peaks_workspace])
+
+
+def remove_noninteger(M):
```

Comment:
can we change the variable name to `matrix` instead of `M` for verbosity and to conform to PEP8?